fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateAggRule.java

```diff
@@ -0,0 +1,153 @@
+import com.starrocks.sql.optimizer.operator.OperatorType;
+import com.starrocks.sql.optimizer.operator.logical.LogicalAggregationOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalProjectOperator;
+import com.starrocks.sql.optimizer.operator.pattern.Pattern;
+import com.starrocks.sql.optimizer.operator.scalar.CallOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ConstantOperator;
+import com.starrocks.sql.optimizer.operator.scalar.IsNullPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.rule.RuleType;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+// When a column used in a SQL query's Group By statement has a unique attribute, aggregation can be eliminated,
+// and the LogicalAggregationOperator can be replaced with a LogicalProjectOperator.
+//
+// Pattern:
+//
+//      Agg
+//       |
+//     Child
+//
+// Transform:
+//
+//     Project
+//       |
+//     Child
+//
+//
+// example:
+// 1. select count(row1) from demo group by pk -> SELECT IF(row1 IS NULL, 0, 1) from demo;
+// 2. select avg(row1) from demo group by pk -> SELECT row1 FROM demo;
+//
+
+public class EliminateAggRule extends TransformationRule {
+
+    private EliminateAggRule() {
+        super(RuleType.TF_ELIMINATE_AGG, Pattern.create(OperatorType.LOGICAL_AGGR)
+                .addChildren(Pattern.create(OperatorType.PATTERN_LEAF)));
+    }
+
+    public static EliminateAggRule getInstance() {
+        return INSTANCE;
+    }
+
+    private static final EliminateAggRule INSTANCE = new EliminateAggRule();
+
+    @Override
+    public boolean check(OptExpression input, OptimizerContext context) {
+        LogicalAggregationOperator aggOp = input.getOp().cast();
+        if (aggOp.getGroupingKeys().isEmpty()) {
+            return false;
+        }
+
+        for (Map.Entry<ColumnRefOperator, CallOperator> entry : aggOp.getAggregations().entrySet()) {
+            String fnName = entry.getValue().getFnName();
+            if (!(fnName.equals(FunctionSet.SUM) || fnName.equals(FunctionSet.COUNT) ||
+                    fnName.equals(FunctionSet.AVG) ||
+                    fnName.equals(FunctionSet.FIRST_VALUE) ||
+                    fnName.equals(FunctionSet.MAX) || fnName.equals(FunctionSet.MIN) ||
+                    fnName.equals(FunctionSet.GROUP_CONCAT))) {
+                return false;
+            }
+        }
+
+        // collect uk pk key
+        UKFKConstraintsCollector collector = new UKFKConstraintsCollector();
+        input.getOp().accept(collector, input, null);
+
+        OptExpression childOptExpression = input.inputAt(0);
+        for (ColumnRefOperator columnRefOperator : aggOp.getGroupingKeys()) {
+            if (childOptExpression.getConstraints().getUniqueConstraint(columnRefOperator.getId()) == null) {
```

Comment:
the rule only takes effect only when group by keys contain all unique key. If group by key is (colA, colB) but unique key is (colA, colB, colC), we cannot remove the agg.